### PR TITLE
force aliases in deparsing for queries with anonymous column references

### PR DESCRIPTION
**Pull request by thanodnl**
DESCRIPTION: Force aliases in deparsing for queries with anonymous column references

Fixes: #3985  

The root cause has todo with discrepancies in the query tree we create. I think in the future we should spend some time on categorising all changes we made to ruleutils and see if we can change the data structure `query` we pass to the deparser to have an actual valid postgres query for the deparser to render.

For now the fix is to keep track, besides changing the names of the entries in the target list, also if we have a reference to an anonymous columns. If there are anonymous columns we set the `printaliases` flag to true which forces the deparser to add the aliases.

TODO
 - [x] Find more test cases, eg `?column?` as a table column name.

### CODE REVIEW

File: src/backend/distributed/deparser/ruleutils_11.c [hunk 1]
```diff
@@ -1053,6 +1054,7 @@ set_relation_column_names(deparse_namespace *dpns, RangeTblEntry *rte,
 	 */
 	noldcolumns = list_length(rte->eref->colnames);
 	changed_any = false;
+	has_anonymous = false;
```
Comment (marcocitus) [inline src/backend/distributed/deparser/ruleutils_11.c:1057]: maybe `has_anonymous_column`

File: src/test/regress/expected/anonymous_columns.out [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+(1 row)
+
+INSERT INTO t0 VALUES (1, 2, 'hello'), (2, 4, 'world');
+SELECT "?column?" FROM t0;
+ ?column?
+---------------------------------------------------------------------
+ hello
+ world
+(2 rows)
+
```
Comment (marcocitus) [inline src/test/regress/expected/anonymous_columns.out:18]: maybe add 
```
WITH a AS (SELECT 1, * FROM t0 WHERE a = 1) SELECT * FROM a;
```

File: src/backend/distributed/deparser/ruleutils_11.c [hunk 3]
```diff
@@ -1090,6 +1092,10 @@ set_relation_column_names(deparse_namespace *dpns, RangeTblEntry *rte,
 		/* Remember if any assigned aliases differ from "real" name */
 		if (!changed_any && strcmp(colname, real_colname) != 0)
 			changed_any = true;
+
+		/* Remember if there is a reference to an anonymous column */
```
[the remark this answers is not in the document]
Comment (SaitTalhaNisanci) [reply src/backend/distributed/deparser/ruleutils_11.c:1096]:
```c
/*
 * FigureColname -
 *	  if the name of the resulting column is not specified in the target
 *	  list, we have to guess a suitable name.  The SQL spec provides some
 *	  guidance, but not much...
 *
 * Note that the argument is the *untransformed* parse tree for the target
 * item.  This is a shade easier to work with than the transformed tree.
 */
char *
FigureColname(Node *node)
{
	char	   *name = NULL;

	(void) FigureColnameInternal(node, &name);
	if (name != NULL)
		return name;
	/* default result if we can't guess anything */
	return "?column?";
}
```

File: src/test/regress/sql/anonymous_columns.sql [hunk 4]
```diff
@@ -0,0 +1,39 @@
+SET citus.next_shard_id TO 20080000;
```
Comment (SaitTalhaNisanci) [inline src/test/regress/sql/anonymous_columns.sql:1]: would be good to order by the results in this file, they seem to be flaky.

Comment (SaitTalhaNisanci) [reply src/test/regress/sql/anonymous_columns.sql:1]: https://circleci.com/gh/citusdata/citus/141846

**Pull request merged**